src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs

```diff
@@ -69,39 +52,152 @@ protected virtual SyntaxNode GetNodeToAnalyze(SyntaxNode root, TextSpan span)
+            // case 2: This is a nested type, offer to move to new file.
+            // case 3: If there is a single type decl in current file, *do not* offer move to new file,
+            //         rename actions are sufficient in this case.
+            if (manyTypes || isNestedType)
             {
-                // If there are multiple type declarations in current document. offer, move to new file.
-                // Or if this is a nested type, offer to move to new file.
-                actions.Add(GetCodeAction(state, operationKind: OperationKind.MoveType));
+                foreach (var fileName in suggestedFileNames)
+                {
+                    actions.Add(GetCodeAction(state, fileName, operationKind: OperationKind.MoveType));
+                }
             }
-            else
+
+            // (2) Add rename file and rename type code actions:
+            // case 1: Multiple type decls in file and none match file name.
+            // case 2: Single type declaration in file and names don't match.
+            if (!AnyTopLevelTypeMatchesDocumentName(state, cancellationToken))
             {
-                // one type declaration in current document. No moving around required, just sync
-                // document name and type name by offering rename in both directions between type and document.
-                actions.Add(GetCodeAction(state, operationKind: OperationKind.RenameFile));
+                foreach (var fileName in suggestedFileNames)
+                {
+                    actions.Add(GetCodeAction(state, fileName, operationKind: OperationKind.RenameFile));
+                }
 
                 // only if the document name can be legal identifier in the language,
                 // offer to rename type with document name
                 if (state.IsDocumentNameAValidIdentifier)
                 {
-                    actions.Add(GetCodeAction(state, operationKind: OperationKind.RenameType));
+                    actions.Add(GetCodeAction(state, fileName: state.DocumentName, operationKind: OperationKind.RenameType));
                 }
             }
 
             return actions.ToImmutableArray();
         }
 
-        private CodeAction GetCodeAction(
-            State state,
-            OperationKind operationKind)
+        private CodeAction GetCodeAction(State state, string fileName, OperationKind operationKind) =>
+            new MoveTypeCodeAction((TService)this, state, operationKind, fileName);
+
+        private bool IsNestedType(TTypeDeclarationSyntax typeNode) =>
+            typeNode.Parent is TTypeDeclarationSyntax;
+
+        /// <summary>
+        /// checks if there is a single top level type declaration in a document
+        /// </summary>
+        private bool MultipleTopLevelTypeDeclarationInSourceDocument(SyntaxNode root) =>
+            TopLevelTypeDeclarations(root).Count() > 1;
```

Comment:
Slightly faster would be .Skip(1).Any().  I'm a little worried about traversing the entire tree for things like Syntax.xml.